src/core/cluster.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Collection of global cluster state for the ZooKeeper quorum."""
+import logging
+from typing import Dict, Optional, Set
+
+from charms.data_platform_libs.v0.data_interfaces import DataPeerOtherUnit
+from ops.framework import Framework, Object
+from ops.model import Relation, Unit
+
+from core.models import (
+    SUBSTRATES,
+    CharmWithRelationData,
+    ODCluster,
+    ODServer,
+    OpensearchServer,
+)
+from literals import OPENSEARCH_REL_NAME, PEER
+
+logger = logging.getLogger(__name__)
+
+
+class ClusterStateBase(Object):
+    """Collection of global cluster state for Framework/Object."""
+
+    def __init__(self, charm: Framework | Object, substrate: SUBSTRATES):
+        super().__init__(parent=charm, key="charm_state")
+        self.substrate: SUBSTRATES = substrate
+
+    # --- RAW RELATION ---
+
+    @property
+    def peer_relation(self) -> Relation | None:
+        """The cluster peer relation."""
+        return self.model.get_relation(PEER)
+
+    @property
+    def opensearch_relation(self) -> Optional[Relation]:
+        """The Opensearch Server relation."""
+        if self.model.relations.get(OPENSEARCH_REL_NAME):
+            return self.model.relations[OPENSEARCH_REL_NAME][0]
+
+
+class ClusterState(ClusterStateBase):
+    """Collection of global cluster state for the charm."""
+
+    def __init__(self, charm: CharmWithRelationData, substrate: SUBSTRATES):
+        super().__init__(charm, substrate)
+        self.charm = charm
+        self._servers_data = {}
+
+    # --- CORE COMPONENTS---
+
+    @property
+    def unit_server(self) -> ODServer:
+        """The server state of the current running Unit."""
+        return ODServer(
+            relation=self.peer_relation,
+            data_interface=self.charm.peer_unit_interface,
+            component=self.model.unit,
+            substrate=self.substrate,
+        )
+
+    @property
+    def peer_units_data_interfaces(self) -> Dict[Unit, DataPeerOtherUnit]:
+        """The cluster peer relation."""
+        if not self.peer_relation or not self.peer_relation.units:
+            return {}
+
+        for unit in self.peer_relation.units:
+            if unit not in self._servers_data:
```

Comment:
No, we can't. The point is not to re-instantiate a `DataPeerOtherUnit` instance, as it will raise an error.
We can only have one instance per unit. Attempt on second instantiation raises an error.